api.c

```diff
@@ -572,10 +571,6 @@ int tcmu_emulate_evpd_inquiry(
 		/* Optimal xfer length */
 		memcpy(&data[12], &val32, 4);
 
-		/* MAXIMUM WRITE SAME LENGTH */
-		val64 = htobe64(VPD_MAX_WRITE_SAME_LENGTH);
-		memcpy(&data[36], &val64, 8);
-
```

Comment:
With your patch, when we set it to zero, we only get a max of 0xfff, because of this chunk in sd_config_write_same:

`        if (sdkp->max_ws_blocks > SD_MAX_WS10_BLOCKS)
                sdkp->max_ws_blocks = min_not_zero(sdkp->max_ws_blocks,
                                                   (u32)SD_MAX_WS16_BLOCKS);
        else if (sdkp->ws16 || sdkp->ws10 || sdkp->device->no_report_opcodes)
                sdkp->max_ws_blocks = min_not_zero(sdkp->max_ws_blocks,
                                                   (u32)SD_MAX_WS10_BLOCKS);
        else {
                sdkp->device->no_write_same = 1;
                sdkp->max_ws_blocks = 0;
        }`

I had to set it to linux's SD_MAX_WS16_BLOCKS value to get larger WS commands. So with linux you need to report a non zero max for WS10 and WS16. It will then do the right thing. In sd_setup_write_same_cmnd it will then do the right thing WRT lengths of commands and what WS command to use.

I have been testing by checking sysfs and then doing BLKZEROOUT ioctls to /dev/sdXYZ.

For VMware, if you return a zero max length does it to the right thing WRT to the spec?

If you set it to a value greater than 0xffff, does it do something similar to linux where it figures things out?